### ISSUE TITLE
[Feature] Non-Success Type Checking - Responses Union Type

### ISSUE THREAD

**Issue by scottc** [mtennoe/swagger-typescript-codegen#73]
Schema:

```
// ...
"responses": {
  "200": {
    "description": "A Thing.",
    "schema": {
      "$ref": "#/definitions/Thing"
    }
  },
  "400": {
    "description": "Invalid Thing Request.",
    "schema": {
      "type": "string"
    }
  },
  "404": {
    "description": "Thing not found.",
    "schema": {
      "type": "number"
    }
  }
}
// ...
```

Generates typescript response type of:

`Promise < ResponseWithBody < ThingResponse >>`

Expecting:

`Promise < ResponseWithBody < 200ThingResponse | 400ThingResponse | 404ThingResponse >>`

This allows for type checking of non-success responses as well as exhaustive pattern matching.

```
const unreachable = (n: never):any => n;

// pattern match on response type.
const poorMansPatternMatch = (response: ResponseWithBody < 200ThingResponse | 400ThingResponse | 404ThingResponse >): boolean =>
{
switch (response.type) {
   case 200:
      console.log(`Happy object: ${JSON.stringify(response.body)}`);
      return true;

   case 400:
   case 404:
      console.log(`Sad string/number: ${response.body}`);
      return false;

   default:
       unreachable(response.type); // compile time error, for unhandled response body types.
       console.warn(`Unhandled response type: ${response.type}`)
       return false;
}
}
```

**Comment by mtennoe** [comment]
Great idea! Would you be willing to give it a shot?

**Comment by scottc** [comment]
https://swagger.io/docs/specification/2-0/describing-responses/
https://swagger.io/docs/specification/describing-responses/

>  A response is defined by its HTTP status code and the data returned in the response body and/or headers

Another nice addition would be including the HTTP status code information in the type information.

```
Promise < ResponseWithBody < ThingResponseBody | ErrorResponseBody >>
```

I have a few different HTTP status codes that share the ErrorResponseBody type, such as 404, 400.

It's good that i can discriminate on body type, but when i have multiple responses sharing the same body type, it's difficult to discriminate on them.

Would want to also discriminate on them, based on their status code.

```
const errorResponse: ResponseWithBody < 400 ErrorResponseBody > | ResponseWithBody < 404, ErrorResponseBody > = (undefined as any);

switch (response.body.type) {
  case "ERROR":
    break;
  default:
    const n: never = response.status;
}

switch (response.status) {
  case "400":
  case "404":
    break;
  default:
    const n: never = response.status;
}
```

**Comment by mtennoe** [comment]
Definitely makes sense to add. Considering you already have this on a branch, I suggest merging that in with the existing PR

**Comment by scottc** [comment]
Added to pull request.

Found why tests are failing, generating some types that look like this.

```
}): Promise < ResponseWithBody < 200, Authentication > | ResponseWithBody < 401, Error > | ResponseWithBody < security, void > | ResponseWithBody < default, Error >> {
```

I suspect it should be like this instead:

```
}): Promise < ResponseWithBody < 200, Authentication > | ResponseWithBody < 401, Error > > {
```

Fixing...

**Comment by mtennoe** [comment]
Merged the RFC, and published as 1.10.0
